### ISSUE TITLE
Add FOCUSZ into amp headers

### ISSUE THREAD

**Issue by jbkalmbach** [LSSTDESC/imSim#413]
For the AOS LSSTCam Full Array Mode we need to know the defocus distance from the FITS header to figure out which images are intra-focal and which are extra-focal.

**Comment by cwwalter** [comment]
Can you tell us how this will be done in normal operations?  Does this just report the camera hexapod z-position?

**Comment by cwwalter** [comment]
Also note: there is already a YAML command if you just want to add a header.

If we put this in the normal readout, it should be something that will be added in the normal headers in operation, + we need to understand how the value will filled.

**Comment by jbkalmbach** [comment]
> Can you tell us how this will be done in normal operations? Does this just report the camera hexapod z-position?

Let me double check on slack from people who know better but I think it is the user-defined offset in the camera hexapod z-position beyond the corrections from the system.

> Also note: there is already a YAML command if you just want to add a header.

I tried to do this but the value only showed up in the eimage headers not the amp headers. Is there a way to pass it to both?

**Comment by cwwalter** [comment]
> Let me double check on slack from people who know better but I think it is the user-defined offset in the camera hexapod z-position beyond the corrections from the system.

hmm... OK. I was thinking it would be tied to one of the yaml perturbation commands.  Is there a separate one for the user offset?

> I tried to do this but the value only showed up in the eimage headers not the amp headers. Is there a way to pass it to both?

Ah, @jmeyers314 added that option.  Josh, was your intention to only go to the eimage?

**Comment by jbkalmbach** [comment]
> hmm... OK. I was thinking it would be tied to one of the yaml perturbation commands. Is there a separate one for the user offset?

It is when I'm running simulations. It is the same value I add to the aos degrees of freedom:
```
imsimConfigYaml["input"]["telescope"]["fea"]["aos_dof"]["dof"][5] += obsMetadata.focusZ
```

where the degrees of freedom are:
```
    # 0: M2 dz (um)
    # 1,2: M2 dx,dy (um)
    # 3,4: M2 rx,ry (arcsec)
    # 5: camera dz (um)
    # 6,7: camera dx,dy (um)
    # 8,9: camera rx,ry (arcsec)
    # 10-29: M1M3 bending modes (um)
    # 30-49: M2 bending modes (um)
```

**Comment by cwwalter** [comment]
OK. Great. So then we should be able to just set the header to that value in the dictionary if it exists.

Did you confirm that this is a standard header that will normally be in the output?

**Comment by jmeyers314** [comment]
I think it makes sense to add a `focusZ` argument to https://github.com/LSSTDESC/imSim/blob/main/imsim/telescope_loader.py#L122 specifically for intentional defocusing of the camera, and then populate FOCUSZ in both eimages and amp images from there.  This would make FOCUSZ=0 in normal operations, even for WF sensors (@jbkalmbach is that the right convention?)

There's no good way to enforce this, but I think then aos dof [5] should only be a correction factor, not a way to intentionally defocus the telescope.  

Note that this is similar to how the existing `rotTelPos` argument to `load_telescope` works.  Technically, the following are equivalent:

```yaml
input:
  telescope:
    perturbations:
      LSSTCamera:
        rotZ: 30 deg
```

```yaml
input:
  telescope:
    rotTelPos: 30 deg
```

But I think the latter should be preferred for intentionally setting the rotator angle, and the former reserved for something like "simulated rotator errors".

Along those lines I think 

```yaml
input:
  telescope:
    focusZ: 0.0015 # meters
```

should be the preferred way to move the camera extra-focal for the express purpose of making donuts in the science CCDs.

**Comment by jmeyers314** [comment]
>  Josh, was your intention to only go to the eimage?

I don't think I considered what should happen to the amp image headers.  Forwarding things there sounds reasonable.

**Comment by cwwalter** [comment]
This sounds good, but you probably don't want to make them equivalent in this case right? I guess it should be the sum of dof[5] and focus_offset.

**Comment by jmeyers314** [comment]
> This sounds good, but you probably don't want to make them equivalent in this case right? I guess the control parameter should be the sum of dof[5] and focus_offset.

If you populate both, then the net effect should be the sum.

**Comment by jbkalmbach** [comment]
I think Josh's proposal makes a lot of sense and am happy to try and implement it myself. I can close my other PR and try to do this.

> This would make FOCUSZ=0 in normal operations, even for WF sensors (@jbkalmbach is that the right convention?)

That's right.

> Did you confirm that this is a standard header that will normally be in the output?

Yes, just confirmed with Tiago we will have this information in operations.

**Comment by jmeyers314** [comment]
Just a note, after chatting with Tiago, it looks like the value written to the header will include both any 1.5 mm offset to go from in focus to "full array mode", and any additional offset from, e.g., measuring the sensitivity matrix or closing the feedback loop.  I'm still happy to have distinct inputs for focusZ and aos dof [5], but it looks like the value we should write to the header is really the sum of these, and not only the focusz bit.

Note that there's a third way to offset the camera, namely:

```yaml
input:
  telescope:
    perturbations:
      LSSTCamera:
        shift: [0, 0, 0.0015]
```

I'd propose to _not_ also include that in the header FOCUSZ.  We can reserve this perturbation as a way to simulate errors where the value used in the raytracing doesn't precisely match what gets written to the header.

**Comment by jbkalmbach** [comment]
I agree with not including that third way in FOCUSZ. 

I think it would be helpful to still have distinct inputs for focusZ and aos dof [5] because when we set up the FAM simulation we will want to have a manual piston movement that is consistent through each iteration of the closed loop.

**Comment by jbkalmbach** [comment]
Closed by #414
